lib/core/test/unit/Cardano/Pool/DB/Properties.hs

```diff
@@ -822,6 +871,91 @@ prop_listRegisteredPools DBLayer {..} entries =
             ]
         assert (pools == (view #poolId <$> reverse entries))
 
+-- | Test that `listRetiredPools` returns the correct set of retirements for
+--   any given epoch.
+--
+-- This property tests `listRetiredPools` in conditions where:
+--
+--   - there are multiple pools;
+--   - there are multiple registrations and retirements for each pool;
+--   - certificates affecting different pools are interleaved in time.
+--
+prop_listRetiredPools_multiplePools_multipleCerts
+    :: DBLayer IO
+    -> [SinglePoolCertificateSequence]
+    -> ListSerializationMethod
+    -> Property
+prop_listRetiredPools_multiplePools_multipleCerts
+    DBLayer {..} certificateSequences serializationMethod = checkCoverage
+        -- Check the number of certificates:
+        $ cover 2 (certificateCount == 0)
+            "number of certificates: = 0"
+        $ cover 2 (certificateCount > 0 && certificateCount <= 10)
+            "number of certificates: > 0 && <= 10"
+        $ cover 2 (certificateCount > 10 && certificateCount <= 100)
+            "number of certificates: > 10 && <= 100"
+        $ cover 2 (certificateCount > 100 && certificateCount <= 1000)
+            "number of certificates: > 100 && <= 1000"
+        -- Check the number of pools:
+        $ cover 2 (poolCount == 0)
+            "number of pools: = 0"
+        $ cover 2 (poolCount > 0 && poolCount <= 10)
+            "number of pools: > 0 && <= 10"
+        $ cover 2 (poolCount > 10 && poolCount <= 100)
+            "number of pools: > 10 && <= 100"
+        $ monadicIO (setup >> prop)
+  where
+    setup = run $ atomically cleanDB
+
+    prop = do
+        run $ atomically $ do
+            mapM_ (uncurry putCertificate) allPublicationsSerialized
+        lifeCycleStatuses <- run $ atomically $ do
+            mapM readPoolLifeCycleStatus allPoolIds
+        let poolsMarkedToRetire = catMaybes $
+                getPoolRetirementCertificate <$> lifeCycleStatuses
```

Comment:
I'd be better not to rely on `lifeCycleStatuses` to determine which pools are supposed to be retired. The risk is to have a similar logic bug in both `lifeCycleStatuses` and `listRetiredPools` that remain uncaught, especially if the same approach is used behind the scene.

Reply:
> I'd be better not to rely on `lifeCycleStatuses` to determine which pools are supposed to be retired. The risk is to have a similar logic bug in both `lifeCycleStatuses` and `listRetiredPools` that remain uncaught, especially if the same approach is used behind the scene.

I definitely agree, this is indeed a risk.

To counteract this risk:

1. We do have a suite of tests specifically aimed at `determinePoolLifeCycleStatus`:
    - `prop_determinePoolLifeCycleStatus_orderCorrect`
    - `prop_determinePoolLifeCycleStatus_neverRegistered`
    - `prop_determinePoolLifeCycleStatus_differentPools`

2. The definition of `listRetiredPools` uses a completely different approach in the form of a handwritten query. In the property test for this query, we check it against the `determinePoolLifeCycleStatus` function, which we consider to be the source of truth.